query/src/sql/planner/metadata.rs

```diff
@@ -138,6 +138,12 @@ impl Metadata {
         column_index
     }
 
+    pub fn nullable_data_type(&mut self, index: IndexType) {
```

Comment:
We need to generate a new `ColumnEntry`. This may cause the data type of a not null column is mistake changed to nullable.

For example:
```sql
select t1.number from numbers(1) t  left join (select * from numbers(1) where number is null) t1 using(number)
```

The `t1.number` should be not null inside subquery `t1`. But after applying this patch, it'll be changed to nullable, which may potentially bring some bugs.

Reply:
> We need to generate a new `ColumnEntry`

If we add a new `ColumnEntry`, how to decide the new `ColumnEntry`'s name? I think it can't be the same as the old name, otherwise it'll be result in name ambiguous

Reply:
> I think it can't be the same as the old name, otherwise it'll be result in name ambiguous

I think it's fine to use the same name, since the outer join operator will eliminate old columns and produce new columns instead. It won't cause ambiguity.

You can do this as:
1. Check if the data type of a column changes after outer join. If it doesn't,  just pass it through outer join. If it does, go to 2.
2. Use `Metadata::add_column` to generate a new `ColumnEntry` for the column with original name and changed data type. Go to 3
3. Replace the column bindings of the inner table(e.g. right table in `LEFT JOIN`) in output `BindContext` with the column entries generated in step 2.

Reply:
I see. IMO, we can store the output columns inside `OuterJoin` struct, and produce the corresponding output schema with hash join processor.

But whatever, it's time to refactor the executor, let's just merge this PR with your patch.